test/types/community/collection/recursive-types.test-d.ts

```diff
@@ -1,23 +1,70 @@
-import { expectError } from 'tsd';
+import { expectAssignable, expectError, expectNotAssignable, expectNotType } from 'tsd';
 
-import type { Collection } from '../../../../src';
+import type { Collection, Filter, UpdateFilter } from '../../../../src';
 
 /**
  * mutually recursive types are not supported and will not get type safety
  */
-interface A {
-  b: B;
+interface Author {
+  name: string;
+  favoritePublication: Book;
 }
 
-interface B {
-  a: A;
+interface Book {
+  title: string;
+  author: Author;
 }
 
-declare const mutuallyRecursive: Collection<A>;
-//@ts-expect-error
-mutuallyRecursive.find({});
-mutuallyRecursive.find({
-  b: {}
+expectAssignable<Filter<Author>>({
+  favoritePublication: {
+    title: 'book title',
+    author: {
+      name: 'author name'
+    }
+  }
+});
+expectNotType<UpdateFilter<Author>>({
+  $set: {
+    favoritePublication: {
+      title: 'a title',
+      published: new Date(),
+      author: {
+        name: 23
+      }
+    }
+  }
+});
+
+// Extremely deep type checking for recursive schemas
+expectNotAssignable<Filter<Author>>({
+  'favoritePublication.author.favoritePublication.author.favoritePublication.author.favoritePublication.title': 23
```

Comment:
can we note what the expected depth is that we're checking? because right now the test lengths seem arbitrary, like if the idea is to show every depth up to 10, then maybe we should number/present them in increasing order; otherwise going from 8 to 10 to 7 makes it hard to tell what the intention really is and makes it easy to introduce mistakes

for example, I am pretty sure L43 is supposed to check 8 and not 10, because at 10 it passes just because it's beyond the limit

and then in the one that checks 7 (L46), we should include a comparable passing example to demonstrate that it's indeed the type of the property that is the differentiating factor

Reply:
I added a test for every level up to 8 for both Filter and UpdateFilter, should make it clear what works and what doesn't. Thoughts?

Reply:
much easier to read!